File: src/main/java/me/iankit/simplechat/chat/ChatHandler.java
package me.iankit.simplechat.chat;

import com.fasterxml.jackson.core.JsonProcessingException;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.socket.WebSocketHandler;
import org.springframework.web.reactive.socket.WebSocketMessage;
import org.springframework.web.reactive.socket.WebSocketSession;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Component
public class ChatHandler implements WebSocketHandler {
    private final ChatDao chatDao;
    private final GroupDao groupDao;

    //    private final Flux<String> intervalFlux = Flux.interval(Duration.ofMillis(1000L))
//            .zipWith(eventFlux, (time, event) -> event);
    private final ChatMapper chatMapper;

    public ChatHandler(ChatDao chatDao, GroupDao groupDao, ChatMapper chatMapper) {
        this.chatDao = chatDao;
        this.groupDao = groupDao;
        this.chatMapper = chatMapper;
    }

    @Override
    public Mono<Void> handle(WebSocketSession session) {
        Flux<WebSocketMessage> output = session.receive()
                .flatMap(msg -> {
                    try {
                        var action = chatMapper.mapToChatAction(msg);
                        System.out.println("New Action :: " + action.getActionType());

                        switch (action.getActionType()) {
                            case ENTER_GROUP: {
                                var groupId = ((EnterGroupAction) action).getGroupId();
                                System.out.println("Enter Group " + groupId);
                                return chatDao.findByGroupId(groupId);
                            }
                            case SEND_MESSAGE: {
                                var chat = ((SendMessageAction) action).toChat();
                                return chatDao.save(chat).flatMap((c) -> Mono.empty());
                            }
                        }
                        return Mono.empty();
                    } catch (JsonProcessingException e) {
                        System.out.println(e);
                        throw new RuntimeException(e);
                    }
                })
                .map(chatMapper::stringify)
                .log()
                .map(session::textMessage);

        return session.send(output);
    }
}

File: src/main/java/me/iankit/simplechat/chat/ChatHandler.kt
package me.iankit.simplechat.chat

import com.fasterxml.jackson.core.JsonProcessingException
import org.springframework.stereotype.Component
import org.springframework.web.reactive.socket.WebSocketHandler
import org.springframework.web.reactive.socket.WebSocketMessage
import org.springframework.web.reactive.socket.WebSocketSession
import reactor.core.publisher.Mono

@Component
class ChatHandler(
    private val chatDao: ChatDao,
    private val groupDao: GroupDao,
    private val chatMapper: ChatMapper,
) : WebSocketHandler {
    override fun handle(session: WebSocketSession): Mono<Void> {
        println("Chat connected")
        val output = session.receive()
            .flatMap { msg: WebSocketMessage ->
                try {
                    val action = chatMapper.mapToChatAction(msg)
                    println("New Action :: " + action.actionType)

                    when (action) {
                        is EnterGroupAction -> return@flatMap chatDao.findByGroupId(action.groupId)

                        is SendMessageAction -> return@flatMap chatDao.save(action.toChat())
                            .flatMap { Mono.empty<Chat>() }

                        else -> return@flatMap Mono.empty()
                    }
                } catch (e: JsonProcessingException) {
                    println(e)
                    throw RuntimeException(e)
                } catch (e: Exception) {
                    print(e)
                    throw e
                }
            }
            .map(chatMapper::stringify)
            .doOnError { println(it) }
            .map(session::textMessage)
        return session.send(output)
    }
}